CONTRIBUTING.md

```diff
@@ -0,0 +1,91 @@
+- [azure-functions-codegen](https://github.com/peterhuene/azure-functions-rs/tree/master/azure-functions-codegen) - The `azure-functions-codegen` crate that defines the procedural macros that are used when writing Azure Functions with Rust.
+- [azure-functions-sdk](https://github.com/peterhuene/azure-functions-rs/tree/master/azure-functions-sdk) - The `azure-functions-sdk` crate that implements the `cargo func` command.
+- [azure-functions-shared](https://github.com/peterhuene/azure-functions-rs/tree/master/azure-functions-shared) - The `azure-functions-shared` crate that defines types and functions that are shared between the `azure-functions-codegen` and `azure-functions` crates.
+  - Note: the `azure-functions-shared/protobuf` directory is the git submodule for [Azure Functions Language Worker Protocol](https://github.com/Azure/azure-functions-language-worker-protobuf).
+- [azure-functions-shared-codegen](https://github.com/peterhuene/azure-functions-rs/tree/master/azure-functions-shared-codegen) - The `azure-functions-shared-codegen` crate that defines the procedural macros used by the shared `azure-functions-shared` crate.
+- [examples](https://github.com/peterhuene/azure-functions-rs/tree/master/examples) - The directory containing example Azure Functions.
+
+## Setting Up a Dev Environment
+
+### Cloning the Repository
+
+This repository uses a git submodule for defining the [Azure Functions Language Worker Protocol](https://github.com/Azure/azure-functions-language-worker-protobuf).
+
+Use `--recurse-submodules` when cloning this repository:
+
+Cloning with SSH:
+
+``` bash
+git clone --recurse-submodules git@github.com:<GITHUB-USERNAME>/azure-functions-rs.git
+```
+
+Cloning with HTTPS:
+
+``` bash
+git clone --recurse-submodules https://github.com/<GITHUB-USERNAME>/azure-functions-rs.git
+```
+
+If you want to clone the source repository, replace `<GITHUB-USERNAME>` with `peterhuene`. To clone your own fork, replace the value with your GitHub username.
+
+### Building
+
+Build at the root of the repository to build both the `azure-functions-codegen` and the `azure-functions` libraries using `cargo build`:
+
+``` bash
+cargo build
+```
+
+### Running tests
+
+Use `cargo test` to run the tests:
+
+``` bash
+cargo test
+```
+
+### Updating Your Fork
+
+As you work on your contributions, code on the source repository may get updated. You can keep your fork up-to-date and avoid merge conflict by adding the source repo as a remote, upstream branch.
+
+``` bash
+git remote add upstream https://github.com/peterhuene/azure-functions-rs
+```
+
+You only need to do this once. Then, to update your fork fetch all branches of the source repo
+
+``` bash
+git fetch upstream
+```
+
+Switch to `master` and rebase so that any commits you have made which aren't in upstream/master are replaced on top of that other branch:
```

Comment:
Ideally, contributors aren't ever making commits to their master/dev branches, and instead using a feature branch which gets merged upstream and then they can pull those commits into their local master/dev branches.

I generally use `git pull --rebase upstream <branch>` to pull upstream changes into my feature branches as a single command.

Reply:
Cool, I will edit to reflect that.

Reply:
How is this? Am I thinking about updating correctly?

### Developing Your Contribution

Create a new branch from `dev` for the feature you are adding on your fork. Name the branch according to the feature you are adding (i.e. `binding-to-iot`). As you work on your contributions, code on the source repository may get updated. You can keep your fork up-to-date and avoid merge conflicts by adding the source repo as a remote, upstream branch.

``` bash
git remote add upstream https://github.com/peterhuene/azure-functions-rs
```

You only need to do this once. Then, to update your fork run

``` bash
git pull --rebase upstream <YOUR-FEATURE-BRANCH>
```

This will update your code with the changes that have occurred in the source repo.

Reply:
I like it!